### DISCUSSION TITLE
Allow remote files to be included via include-before-*

### DISCUSSION THREAD

**Discussion by jbullock35** [quarto-dev/quarto-cli#6854] [Q&A]
### Description

Quarto 1.3.450 can't handle a YAML header like this:
```
---
title: "My Quarto document"
format:
  html:
    include-before-body: https://example.com/myMathjax.js
---
```
The problem is that the argument to `include-before-body` is remote. My understanding is that pandoc itself can't handle remote files as arguments to the `include-before` options, and that this is why Quarto chokes in such cases.

But R Markdown *can* handle remote includes. For example, this YAML header poses no problems to R Markdown:
```
---
title: "My R Markdown document"
output:
  html_document:
    includes:
      before_body: https://example.com/myMathjax.js
---
```
I gather that the remote include works for R Markdown because knitr downloads the remote file before passing it to pandoc. Can something similar be done for Quarto, so that Quarto can handle remote files?

**Comment by mcanouil** [comment]
The idea with Quarto is really focused on making the file/project reproducible and self-contained.
There are already few discussions about this kind of feature request.
If you have resources like this, the way to go is to create your extension format which has the proper include.

**Reply by jbullock35** [reply]
I love the goals of making files reproducible and self-contained—and allowing for remote includes would advance those goals. Remote includes could be contained in the output HTML document, as they are with R Markdown. And reproducibility would be enhanced because others would be able to render your QMD file even if they didn't have a local copy of the include file.

Am I missing something? I didn't find any of the discussions to which you allude, but if there are that seem especially relevant, please let me know.

**Reply by mcanouil** [reply]
Nothing would check that remote file changed or not, how is that reproducible?

Related thread:

- https://github.com/quarto-dev/quarto-cli/discussions/5043

For short, if one element is allowed to be remote, why not everything, if everything is remote, how to ensure you can rebuild a document/project and get the exact same results every time using the same version of Quarto?

**Reply by mcanouil** [reply]
By the way, you can leverage the projects script to retrieve remote content yourself, see <https://quarto.org/docs/projects/scripts.html#pre-and-post-render>.
This is the responsibility of reproducibility falls on you and depending on your goal it might be enough.

**Reply by jbullock35** [reply]
Yes, I take your point. But how does limiting the `include-before-body` option (or any other include option) to local files do any better? After all, there's no check that the local file has changed, either. If you want someone else to render your QMD file and get the same results, you'll need to share both your QMD file and your include file with him. This process gets a little easier if your include file is remote—in that case, you need to share only the QMD file.

Of course, if you use the `self-contained` option, your rendered HTML file can be shared and displayed by anyone. It won't change even if the include file changes. But that's true whether the include file is local or remote.

**Reply by mcanouil** [reply]
You don't have to worry about file changes if you share Git repository.

Sharing only part of the source code is not sharing the source code.

Also note that the `self-contained` option was deprecated a year ago, the correct option is `embed-resources`, but don't make your source code reproducible. It just make the rendered document more portable which has nothing to do with reproducibility.

**Reply by jbullock35** [reply]
> You don't have to worry about file changes if you share Git repository.

Yes. I've been bearing in mind the case—sadly common in academia—in which collaborators don't know how to use version control and aren't going to learn.

Thanks again for all of your help.

**Reply by mcanouil** [reply]
> in which collaborators don't know how to use version control and aren't going to learn.

Coming from academia, even before that, to contribute to a Quarto document they are going to learn Markdown or use an IDE compatible with Quarto Visual Editor, so the issue is less about remote files, etc. than teach collaborators to work with Quarto itself.

**Comment by jimjam-slam** [comment]
@jbullock35 I believe [the Includes options](https://quarto.org/docs/reference/formats/pdf.html#includes) can all take either `file` key or a `text` key. So could you try something like the following?

```yaml
---
title: "My Quarto document"
format:
  html:
    include-before-body:
      text: |
        <script src="https://example.com/myMathjax.js"></script>
---
```
